File: commons/src/main/java/org/dhis2/commons/data/tuples/Quartet.java
package org.dhis2.commons.data.tuples;

import androidx.annotation.NonNull;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class Quartet<A, B, C, D> {

    @NonNull
    public abstract A val0();

    @NonNull
    public abstract B val1();

    @NonNull
    public abstract C val2();

    @NonNull
    public abstract D val3();

    @NonNull
    public static <A, B, C, D> Quartet<A, B, C, D> create(@NonNull A val0, @NonNull B val1,
                                                          @NonNull C val2, @NonNull D val3) {
        return new AutoValue_Quartet<>(val0, val1, val2, val3);
    }
}

File: app/src/main/java/org/dhis2/data/tuples/Quartet.kt
package org.dhis2.data.tuples

import com.google.auto.value.AutoValue

@AutoValue
abstract class Quartet<A, B, C, D> {

    abstract fun val0(): A

    abstract fun val1(): B

    abstract fun val2(): C

    abstract fun val3(): D

    companion object {

        fun <A, B, C, D> create(val0: A, val1: B,
                                val2: C, val3: D): Quartet<A, B, C, D> {
            return AutoValue_Quartet(val0, val1, val2, val3)
        }
    }
}
